File: server/web_server/nginx/install_nginx_http_proxy.py
#!/usr/bin/env python3

import subprocess
import os

# Function to clear the screen
def display_clean():
    os.system('clear')
    print("")

# Function to run commands and handle errors
def run_command(command):
    result = subprocess.run(command, shell=True, capture_output=True, text=True)
    if result.returncode != 0:
        print(f"Command failed: {command}")
        print(result.stderr)
        exit(1)

# Function to check if Nginx is installed
def check_nginx_installed():
    result = subprocess.run("pkg info nginx", shell=True, capture_output=True, text=True)
    if result.returncode == 0:
        print("Nginx is already installed on this system.")
        while True:
            reinstall_choice = input("Do you want to reinstall Nginx? (y/n): ").strip().lower()
            if reinstall_choice == 'y':
                run_command("pkg remove -y nginx")
                run_command("pkg install -y nginx")
                return True
            elif reinstall_choice == 'n':
                print("Nginx reinstallation canceled.")
                return False
            else:
                print("Invalid input. Please enter 'y' for yes or 'n' for no.")
    return True

# Function to check Nginx status and start it if it's not running
def start_nginx_if_not_running():
    result = subprocess.run("service nginx status", shell=True, capture_output=True, text=True)
    if result.returncode == 0:
        print("Nginx is already running.")
    else:
        print("Starting Nginx...")
        run_command("service nginx start")

# Function to install Nginx
def install_nginx():
    print("Installing Nginx HTTP Server...")
    run_command("pkg install -y nginx")
    run_command("sysrc nginx_enable=YES")
    start_nginx_if_not_running()

# Function to configure Nginx as HTTP/Proxy server
def configure_nginx():
    nginx_conf = "/usr/local/etc/nginx/nginx.conf"
    print("Configuring Nginx as an HTTP/Proxy server...")

    # Create a backup of the original configuration file
    run_command(f"cp {nginx_conf} {nginx_conf}.bak")

    # Write new configuration to the nginx.conf file
    nginx_configuration = """
user  www;
worker_processes  auto;
error_log  /var/log/nginx/error.log warn;
pid        /var/run/nginx.pid;

events {
    worker_connections  1024;
}

http {
    include       mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;
    keepalive_timeout  65;

    # Simple HTTP server
    server {
        listen       80;
        server_name  localhost;

        location / {
            root   /usr/local/www/nginx;
            index  index.html index.htm;
        }

        error_page  500 502 503 504  /50x.html;
        location = /50x.html {
            root   /usr/local/www/nginx-dist;
        }
    }

    # Proxy server
    server {
        listen 8080;

        location / {
            proxy_pass http://127.0.0.1:80;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }
    }
}
"""
    with open(nginx_conf, 'w') as f:
        f.write(nginx_configuration)
    
    print("Nginx configuration updated.")
    run_command("service nginx restart")

def main():
    display_clean()

    print("This script will install and configure Nginx as an HTTP and Proxy server.")

    # Check if Nginx is installed
    if check_nginx_installed():
        print("Proceeding with the installation and configuration of Nginx.")

        # Install Nginx
        install_nginx()

        # Configure Nginx
        configure_nginx()

        print("")
        print("             ,        ,")
        print("            /(        )`")
        print("            \\ \\___   / |")
        print("            /- _  `-/  '")
        print("           (/\\/ \\ \\   /\\")
        print("           / /   | `    \\")
        print("           O O   ) /    |")
        print("           `-^--'`<     '")
        print("          (_.)  _  )   /")
        print("           `.___/   /")
        print("             `-----' /")
        print("        <----.     __\\ ")
        print("        <----|====O)))==)")
        print("        <----'    `--'")
        print("             `-----'")
        print("     ")
        print("")

        print("Nginx installation and configuration completed.")
    else:
        print("Nginx installation is not required.")

if __name__ == "__main__":
    main()

File: server/web_server/nginx/install_nginx_http_proxy.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Function to clear the screen
void display_clean() {
    system("clear");
    printf("\n");
}

// Function to run a system command and handle errors
void run_command(const char *command) {
    int ret = system(command);
    if (ret != 0) {
        printf("Command failed: %s\n", command);
        exit(1);
    }
}

// Function to check if Nginx is installed
int check_nginx_installed() {
    int ret = system("pkg info nginx >/dev/null 2>&1");
    if (ret == 0) {
        char reinstall_choice[10];
        printf("Nginx is already installed on this system.\n");
        while (1) {
            printf("Do you want to reinstall Nginx? (y/n): ");
            scanf("%s", reinstall_choice);
            if (strcmp(reinstall_choice, "y") == 0 || strcmp(reinstall_choice, "Y") == 0) {
                run_command("pkg remove -y nginx");
                run_command("pkg install -y nginx");
                return 1;
            } else if (strcmp(reinstall_choice, "n") == 0 || strcmp(reinstall_choice, "N") == 0) {
                printf("Nginx reinstallation canceled.\n");
                return 0;
            } else {
                printf("Invalid input. Please enter 'y' for yes or 'n' for no.\n");
            }
        }
    }
    return 1;
}

// Function to check Nginx status and start it if it's not running
void start_nginx_if_not_running() {
    int ret = system("service nginx status >/dev/null 2>&1");
    if (ret == 0) {
        printf("Nginx is already running.\n");
    } else {
        printf("Starting Nginx...\n");
        run_command("service nginx start");
    }
}

// Function to install Nginx
void install_nginx() {
    printf("Installing Nginx HTTP Server...\n");
    run_command("pkg install -y nginx");
    run_command("sysrc nginx_enable=YES");
    start_nginx_if_not_running();
}

// Function to configure Nginx as an HTTP/Proxy server
void configure_nginx() {
    const char *nginx_conf = "/usr/local/etc/nginx/nginx.conf";
    printf("Configuring Nginx as HTTP/Proxy server...\n");

    // Create a backup of the original configuration file
    run_command("cp /usr/local/etc/nginx/nginx.conf /usr/local/etc/nginx/nginx.conf.bak");

    // Write new configuration to the nginx.conf file
    FILE *f = fopen(nginx_conf, "w");
    if (f == NULL) {
        printf("Failed to open %s for writing.\n", nginx_conf);
        exit(1);
    }

    fprintf(f,
            "user  www;\n"
            "worker_processes  auto;\n"
            "error_log  /var/log/nginx/error.log warn;\n"
            "pid        /var/run/nginx.pid;\n\n"
            "events {\n"
            "    worker_connections  1024;\n"
            "}\n\n"
            "http {\n"
            "    include       mime.types;\n"
            "    default_type  application/octet-stream;\n\n"
            "    log_format  main  '$remote_addr - $remote_user [$time_local] \"$request\" '\n"
            "                      '$status $body_bytes_sent \"$http_referer\" '\n"
            "                      '\"$http_user_agent\" \"$http_x_forwarded_for\"';\n\n"
            "    access_log  /var/log/nginx/access.log  main;\n\n"
            "    sendfile        on;\n"
            "    keepalive_timeout  65;\n\n"
            "    server {\n"
            "        listen       80;\n"
            "        server_name  localhost;\n\n"
            "        location / {\n"
            "            root   /usr/local/www/nginx;\n"
            "            index  index.html index.htm;\n"
            "        }\n\n"
            "        error_page  500 502 503 504  /50x.html;\n"
            "        location = /50x.html {\n"
            "            root   /usr/local/www/nginx-dist;\n"
            "        }\n"
            "    }\n\n"
            "    server {\n"
            "        listen 8080;\n"
            "        location / {\n"
            "            proxy_pass http://127.0.0.1:80;\n"
            "            proxy_set_header Host $host;\n"
            "            proxy_set_header X-Real-IP $remote_addr;\n"
            "            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;\n"
            "            proxy_set_header X-Forwarded-Proto $scheme;\n"
            "        }\n"
            "    }\n"
            "}\n");

    fclose(f);
    printf("Nginx configuration updated.\n");

    // Restart Nginx to apply the new configuration
    run_command("service nginx restart");
}

int main() {
    display_clean();

    printf("This script will install and configure Nginx as an HTTP and Proxy server.\n");

    // Check if Nginx is installed
    if (check_nginx_installed()) {
        printf("Proceeding with the installation and configuration of Nginx.\n");

        // Install Nginx
        install_nginx();

        // Configure Nginx
        configure_nginx();

        printf("\n"
               "             ,        ,\n"
               "            /(        )`\n"
               "            \\ \\___   / |\n"
               "            /- _  `-/  '\n"
               "           (/\\/ \\ \\   /\\\n"
               "           / /   | `    \\\n"
               "           O O   ) /    |\n"
               "           `-^--'`<     '\n"
               "          (_.)  _  )   /\n"
               "           `.___/   /\n"
               "             `-----' /\n"
               "        <----.     __\\ \n"
               "        <----|====O)))==)\n"
               "        <----'    `--'\n"
               "             `-----'\n"
               "     \n");

        printf("Nginx installation and configuration completed.\n");
    } else {
        printf("Nginx installation is not required.\n");
    }

    return 0;
}

// Run:
// 1.  cc -o install_nginx_http_proxy install_nginx_http_proxy.c
// 2   sudo ./install_nginx_http_proxy
